micro_g/scripts/pose_servoing_controller.py

```diff
@@ -0,0 +1,194 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import argparse
+import sys
+import time
+
+import numpy as np
+import rclpy
+import tf2_ros
+import transforms3d
+from geometry_msgs.msg import PoseStamped
+from interbotix_common_modules.angle_manipulation import angle_manipulation as ang
+from interbotix_xs_modules.xs_robot.arm import InterbotixManipulatorXS
+from rclpy.utilities import remove_ros_args
+from tf2_geometry_msgs import do_transform_pose
+
+
+class XSArmPoseServoingController(InterbotixManipulatorXS):
+    """
+    Tracks a published end effector pose.
+    """
+
+    def __init__(
+        self,
+        robot_model: str,
+        robot_name: str,
+        control_update_rate: float = 10.0,
+        moving_time: float = 0.5,
+        xs_args=None,
+    ):
+        """
+        Create the controller
+
+        Args:
+            robot_model: the robot model
+            robot_name: the robot name
+            control_update_rate: the rate at which the controller is updated (Hz).
+            moving_time: the duration in which all movements should take place (s).
+            xs_args: Extra arguments to be passed to InterbotixManipulatorXS. Defaults
+                to None.
+        """
+        # Initialize the interface to the arm
+        InterbotixManipulatorXS.__init__(
+            self,
+            robot_model=robot_model,
+            robot_name=robot_name,
+            start_on_init=True,
+            moving_time=moving_time,
+            accel_time=moving_time / 2.0,
+            args=xs_args,
+            node_name="pose_servoing",
+        )
+
+        # Set the controller update rate
+        self.rate = self.core.create_rate(control_update_rate)
+
+        # Define the transforms we'll use. Notation is T_{to}{from}.
+        # Frames are:
+        #   - s = space frame, fixed to the robot's base link
+        #   - y = yaw frame, same as s but rotated about z by the current waist angle
+        #   - b = body frame, fixed to the end effector
+        #   - d = desired frame for the end effector
+        self.T_sy = np.identity(4)
+        self.T_sd = self.arm.get_ee_pose_command()
+        self.update_frames()
+
+        # Initialize the tf2 buffer and listener
+        self.tf_buffer = tf2_ros.Buffer()
+        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self.core)
+        # Set the target frame for transformation
+        self.target_frame = f"{robot_name}/base_link"
+
+        # Create a subscription to the pose that we will track
+        self.core.create_subscription(
+            PoseStamped, "desired_eef_pose", self.desired_pose_callback, 10
```

Comment:
Another thing that I've experienced is that configuring the QoS profiles can be easier when using some of the [provided profiles](https://github.com/ros2/rmw/blob/rolling/rmw/include/rmw/qos_profiles.h) (e.g., `qos_profile_sensor_data`, etc.). It can also be useful to define some common profiles for use across packages in case the standard profiles don't work.

Reply:
I forgot to mention that the profile that I usually default to is `qos_profile_default`:

```python
from rclpy.qos import qos_profile_default
```

Unless there is another profile that makes more sense

Reply:
Done.